ramips: mt7621: add support for Xiaomi Mi Router 4

soikjk opened the pull request:
### Hardware
    Power: 12 VDC, 1.5 A
    Connector type: barrel
    CPU1: MediaTek MT7621A (880 MHz, 4 cores)
    FLA1: 128 MiB (ESMT F59L1G81MA)
    RAM1: 128 MiB (ESMT M15T1G1664A)
    WI1 chip1: MediaTek MT7603EN
    WI1 802dot11 protocols: bgn
    WI1 MIMO config: 2x2:2
    WI1 antenna connector: U.FL
    WI2 chip1: MediaTek MT7612EN
    WI2 802dot11 protocols: an+ac
    WI2 MIMO config: 2x2:2
    WI2 antenna connector: U.FL
    ETH chip1: MediaTek MT7621A
    Switch: MediaTek MT7621A
    LAN speed: 10/100/1000
    LAN ports: 2
    WAN speed: 10/100/1000
    WAN ports: 1
    Default IP address: 192.168.31.1
### TTL Serial
[o] TX
[o] GND
[o] RX
[  ] VCC - Do not connect it

### Flash instructions:
1.Create a simple http server using caddy V1.0
2.set ttl enable 
To enable writing to the console, you must reset to factory settings
Then you see uboot boot, press the keyboard 4 button (enter uboot command line)
If it is not successful, repeat the above operation of restoring the factory settings.
After entering the uboot command line, type:

setenv uart_en 1
saveenv
boot

3.use shell in ttl 
   cd /tmp
   wget http://"your computerip:80"/openwrt-ramips-mt7621-xiaomi_mir4-squashfs-kernel1.bin
   wget http://"your computer ip:80"/openwrt-ramips-mt7621-xiaomi_mir4-squashfs-rootfs0.bin
   mtd write openwrt-ramips-mt7621-xiaomi_mir4-squashfs-kernel1.bin kernel1
   mtd write openwrt-ramips-mt7621-xiaomi_mir4-squashfs-rootfs0.bin rootfs0
   nvram set flag_try_sys1_failed=1
   nvram commit
   reboot
4.login to the router http://192.168.1.1/

soikjk committed:
ramips: mt7621: add support for Xiaomi Mi Router 4

dengqf6 commented on package/boot/uboot-envtools/files/ramips:
```diff
@@ -34,6 +34,7 @@ zbtlink,zbt-wg2626)
 	ubootenv_add_uci_config "/dev/mtd1" "0x0" "0x1000" "0x10000"
 	;;
 xiaomi,mir3p|\
+xiaomi,mir4|\
```
Please keep the alphabetical order

soikjk replied:
updated

dengqf6 commented on target/linux/ramips/dts/mt7621_xiaomi_mir4.dts:
```diff
@@ -0,0 +1,187 @@
+
+&ethernet {
+	mtd-mac-address = <&factory 0xe000>;
+	mediatek,portmap = "lwlll";
```
It should be "llllw" according to your `02_network`

soikjk replied:
updated

dengqf6 commented on target/linux/ramips/mt7621/base-files/etc/board.d/01_leds:
```diff
@@ -72,6 +72,11 @@ xiaomi,mir3g)
+	ucidef_set_led_switch "wan-amber"  "WAN (amber)"  "$boardname:amber:wan"  "switch0" "0x10" "0x08"
+	ucidef_set_led_switch "lan1-amber" "LAN1 (amber)" "$boardname:amber:lan1" "switch0" "0x02" "0x08"
+	ucidef_set_led_switch "lan2-amber" "LAN2 (amber)" "$boardname:amber:lan2" "switch0" "0x04" "0x08"
+	;;
```
Please keep the alphabetical order

adrianschmutzler replied:
And as in the DTS, I do not see a point in adding the color if only one color per LED exists.

dengqf6 commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -138,6 +138,10 @@ ramips_setup_interfaces()
+	xiaomi,mir4)
+		ucidef_add_switch "switch0" \
+			"1:lan:2" "2:lan:1" "4:wan" "6t@eth0"
+		;;
```
ditto

dengqf6 commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -275,6 +279,7 @@ ramips_setup_macs()
 		label_mac=$wan_mac
 		;;
 	xiaomi,mir3g|\
+	xiaomi,mir4|\
```
ditto

dengqf6 commented on target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh:
```diff
@@ -53,6 +53,7 @@ platform_do_upgrade() {
 	netgear,r6850|\
 	netis,wf2881|\
 	xiaomi,mir3g|\
+	xiaomi,mir4|\
```
ditto

dengqf6 commented on target/linux/ramips/dts/mt7621_xiaomi_mir4.dts:
```diff
@@ -0,0 +1,187 @@
+			label = "reset";
+			gpios = <&gpio0 18 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
```
There is another GPIO key on the router named "MiNET"

adrianschmutzler commented:
As with the other mt7621 PRs, I'd like to block this until 5.4 support is merged.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_xiaomi_mir4.dts:
```diff
@@ -0,0 +1,187 @@
+/dts-v1/;
```
Please add a proper SPDX license identifier.

adrianschmutzler commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -860,6 +860,25 @@ define Device/xiaomi_mir3g
 endef
 TARGET_DEVICES += xiaomi_mir3g
 
+define Device/xiaomi_mir4
```
All use correct alphabetic sorting here, i.e. put this below mir3p.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_xiaomi_mir4.dts:
```diff
@@ -0,0 +1,187 @@
+			gpios = <&gpio0 10 GPIO_ACTIVE_LOW>;
+		};
+
+		wan_amber {
```
If there is only one color for a LED type, you can remove the color from the node name:
wan_amber -> wan
Same below for lan1/lan2

dengqf6 replied:
As in this pic ![ports](https://s3.ifanr.com/wp-content/uploads/2018/06/8S5A4611.jpg)

It's green. On MIR3G there are green and amber LEDs, the green one is controlled by the switch and the amber one is controlled by GPIO.

@soikjk Have you tried removing the port LEDs and checking if it still works? If yes, then it's controlled by the switch so you don't need that definition.

adrianschmutzler commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -860,6 +860,25 @@ define Device/xiaomi_mir3g
+  UBINIZE_OPTS := -E 5
+  IMAGES += kernel1.bin rootfs0.bin
+  IMAGE/kernel1.bin := append-kernel
+  IMAGE/rootfs0.bin := append-ubi | check-size
```
That's quite uncommon, isn't it?

soikjk replied:
mir4  and mir3g are basically the same configuration, but without USB interface, RAM 256MiB →128MiB. So I used mir3g's compilation options

adrianschmutzler commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -860,6 +860,25 @@ define Device/xiaomi_mir3g
+  DEVICE_VENDOR := Xiaomi
+  DEVICE_MODEL := Mi Router 4
+  SUPPORTED_DEVICES += R4
+  SUPPORTED_DEVICES += mir4
```
Drop both supported_devices lines.

adrianschmutzler commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -860,6 +860,25 @@ define Device/xiaomi_mir3g
+  SUPPORTED_DEVICES += R4
+  SUPPORTED_DEVICES += mir4
+  DEVICE_PACKAGES := kmod-mt7603 kmod-mt76x2 wpad-basic \
+  uboot-envtools
```
Indent with one tab for this line.

adrianschmutzler commented:
Your PR has formal issues. Please add a proper commit message and Signed-off-by.
https://openwrt.org/submitting-patches

soikjk committed:
ramips: mt7621: add support for Xiaomi Mi Router 4

mir4 is the same as mir3g, except for the RAM(256Mib→
128Mib) LED and gpio(MiNet button).

Specification
=============
Power: 12 VDC, 1.5 A
Connector type: barrel
CPU1: MediaTek MT7621A (880 MHz, 4 cores)
FLA1: 128 MiB (ESMT F59L1G81MA)
RAM1: 128 MiB (ESMT M15T1G1664A)
WI1 chip1: MediaTek MT7603EN
WI1 802dot11 protocols: bgn
WI1 MIMO config: 2x2:2
WI1 antenna connector: U.FL
WI2 chip1: MediaTek MT7612EN
WI2 802dot11 protocols: an+ac
WI2 MIMO config: 2x2:2
WI2 antenna connector: U.FL
ETH chip1: MediaTek MT7621A
Switch: MediaTek MT7621A

TTL Serial
[o] TX
[o] GND
[o] RX
[ ] VCC - Do not connect it

Flash instruction
=================
1.Create a simple http server using caddy V1.0
2.set ttl enable
To enable writing to the console, you must reset to factory settings
Then you see uboot boot, press the keyboard 4 button (enter uboot command line)
If it is not successful, repeat the above operation of restoring the factory settings.
After entering the uboot command line, type:

setenv uart_en 1
saveenv
boot

3.use shell in ttl
cd /tmp
wget http://"your computerip:80"/openwrt-ramips-mt7621-xiaomi_mir4-squashfs-kernel1.bin
wget http://"your computer ip:80"/openwrt-ramips-mt7621-xiaomi_mir4-squashfs-rootfs0.bin
mtd write openwrt-ramips-mt7621-xiaomi_mir4-squashfs-kernel1.bin kernel1
mtd write openwrt-ramips-mt7621-xiaomi_mir4-squashfs-rootfs0.bin rootfs0
nvram set flag_try_sys1_failed=1
nvram commit
reboot
4.login to the router http://192.168.1.1/

Signed-off-by: Zhangshiru <2548379078@qq.com>

mrkiko commented:
Please add your Signed-Off-By tag. sorry but those things are required.

soikjk commented:
> Please add your Signed-Off-By tag. sorry but those things are required.

updated

mrkiko commented:
thank you man! Hoping the best for this commit. thank you for your work.

soikjk committed:
Merge branch 'master' into master

selurvedu commented:
@soikjk are you going to reopen this?

Neustradamus commented:
Good PR is :
- https://github.com/openwrt/openwrt/pull/3649

Merged commit is:
- https://github.com/openwrt/openwrt/commit/c2a7bb520a0fb107984ad3757174d391796cd222
